src/magic.cpp

```diff
@@ -195,3 +195,611 @@ bool spell_type::is_valid() const
+    return type->base_casting_time;
+}
+
+std::string spell::name() const
+{
+    return _( type->name );
+}
+
+float spell::spell_fail( const player &p ) const
+{
+    // formula is based on the following:
+    // exponential curve
+    // effective skill of 0 or less is 100% failure
+    // effective skill of 8 (8 int, 0 spellcraft, 0 spell level, spell difficulty 0) is ~50% failure
+    // effective skill of 30 is 0% failure
+    const float effective_skill = 2 * ( get_level() - get_difficulty() ) + p.get_int() +
+                                  p.get_skill_level( skill_id( "SPELLCRAFT" ) );
+    // add an if statement in here because sufficiently large numbers will definitely overflow because of exponents
+    if( effective_skill > 30.0f ) {
+        return 0.0f;
+    } else if( effective_skill < 0.0f ) {
+        return 1.0f;
+    }
+    const float fail_chance = pow( ( effective_skill - 30.0f ) / 30.0f, 2 );
+    return clamp( fail_chance, 0.0f, 1.0f );
+}
+
+std::string spell::colorized_fail_percent( const player &p ) const
+{
+    const float fail_fl = spell_fail( p ) * 100.0f;
+    std::string fail_str;
+    fail_fl == 100.0f ? fail_str = _( "Difficult!" ) : fail_str = _( string_format( "%.1f %% %s",
+                                   fail_fl, "Failure Chance" ) );
+    nc_color color;
+    if( fail_fl > 90.0f ) {
+        color = c_magenta;
+    } else if( fail_fl > 75.0f ) {
+        color = c_red;
+    } else if( fail_fl > 60.0f ) {
+        color = c_light_red;
+    } else if( fail_fl > 35.0f ) {
+        color = c_yellow;
+    } else if( fail_fl > 15.0f ) {
+        color = c_green;
+    } else {
+        color = c_light_green;
+    }
+    return colorize( fail_str, color );
+}
+
+int spell::xp() const
+{
+    return experience;
+}
+
+void spell::gain_exp( int nxp )
+{
+    experience += nxp;
+}
+
+std::string spell::energy_string() const
+{
+    switch( type->energy_source ) {
+        case hp_energy:
+            return _( "health" );
+        case mana_energy:
+            return _( "mana" );
+        case stamina_energy:
+            return _( "stamina" );
+        case bionic_energy:
+            return _( "bionic power" );
+        default:
+            return "";
+    }
+}
+
+std::string spell::energy_cost_string( const player &p ) const
+{
+    if( energy_source() == none_energy ) {
+        return _( "none" );
+    }
+    if( energy_source() == bionic_energy || energy_source() == mana_energy ) {
+        return colorize( to_string( energy_cost() ), c_light_blue );
+    }
+    if( energy_source() == hp_energy ) {
+        auto pair = get_hp_bar( energy_cost(), p.get_hp_max() / num_hp_parts );
+        return colorize( pair.first, pair.second );
+    }
+    if( energy_source() == stamina_energy ) {
+        auto pair = get_hp_bar( energy_cost(), p.get_stamina_max() );
+        return colorize( pair.first, pair.second );
+    }
+    debugmsg( _( "ERROR: Spell %s has invalid energy source." ), id().c_str() );
```

Comment:
```suggestion
    debugmsg( "ERROR: Spell %s has invalid energy source.", id().c_str() );
```

Reply:
i don't quite follow the reasoning behind these two changes -- my original reasoning behind adding the translation string was that since this is a highly moddable thing that if a modder was making edits and adding spells in another language they might want this error message in their language.

Reply:
Based on #30373 and #30332.